File: mq/main/javax-jms/src/main/java/javax/jms/TextMessage.java
/*
 * Copyright (c) 1997, 2017 Oracle and/or its affiliates. All rights reserved.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0, which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the
 * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
 * version 2 with the GNU Classpath Exception, which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 */

package javax.jms;

/** A {@code TextMessage} object is used to send a message containing a 
  * {@code java.lang.String}.
  * It inherits from the {@code Message} interface and adds a text message 
  * body.
  *
  * <P>This message type can be used to transport text-based messages, including
  *  those with XML content.
  *
  * <P>When a client receives a {@code TextMessage}, it is in read-only 
  * mode. If a client attempts to write to the message at this point, a 
  * {@code MessageNotWriteableException} is thrown. If 
  * {@code clearBody} is 
  * called, the message can now be both read from and written to.
  *
  * @see         javax.jms.Session#createTextMessage()
  * @see         javax.jms.Session#createTextMessage(String)
  * @see         javax.jms.BytesMessage
  * @see         javax.jms.MapMessage
  * @see         javax.jms.Message
  * @see         javax.jms.ObjectMessage
  * @see         javax.jms.StreamMessage
  * @see         java.lang.String
  * 
  * @version JMS 2.0
  * @since JMS 1.0
  * 
  */
 
public interface TextMessage extends Message { 

    /** Sets the string containing this message's data.
      *  
      * @param string the {@code String} containing the message's data
      *  
      * @exception JMSException if the JMS provider fails to set the text due to
      *                         some internal error.
      * @exception MessageNotWriteableException if the message is in read-only 
      *                                         mode.
      */ 

    void
    setText(String string) throws JMSException;


    /** Gets the string containing this message's data.  The default
      * value is null.
      *  
      * @return the {@code String} containing the message's data
      *  
      * @exception JMSException if the JMS provider fails to get the text due to
      *                         some internal error.
      */ 

    String
    getText() throws JMSException;
}

File: mq/src/share/cclient/client/TextMessage.cpp
/*
 * Copyright (c) 2000, 2017 Oracle and/or its affiliates. All rights reserved.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0, which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the
 * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
 * version 2 with the GNU Classpath Exception, which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 */

/*
 * @(#)TextMessage.cpp	1.3 06/26/07
 */ 

#include "TextMessage.hpp"
#include "../util/UtilityMacros.h"
#include "../io/PacketType.hpp"

/*
 *
 */
TextMessage::TextMessage() : Message()
{
  CHECK_OBJECT_VALIDITY();

  if (packet != NULL) {
    packet->setPacketType(PACKET_TYPE_TEXT_MESSAGE);
  }

  messageBodyText = NULL;
}

/*
 *
 */
TextMessage::TextMessage(Packet * const packetArg) : Message(packetArg)
{
  CHECK_OBJECT_VALIDITY();

  ASSERT( packetArg != NULL );
  ASSERT( packetArg->getPacketType() == PACKET_TYPE_TEXT_MESSAGE );

  messageBodyText = NULL;
}

/*
 *
 */
TextMessage::~TextMessage()
{
  CHECK_OBJECT_VALIDITY();

  DELETE( messageBodyText );
}


/*
 *
 */
PRUint16 
TextMessage::getType()
{
  CHECK_OBJECT_VALIDITY();

  return PACKET_TYPE_TEXT_MESSAGE;
}


/*
 *
 */
iMQError
TextMessage::setMessageText(const UTF8String * const messageText)
{
  CHECK_OBJECT_VALIDITY();

  iMQError errorCode = IMQ_SUCCESS;
  const PRUint8 * textBytes = NULL;
  PRInt32 textBytesSize = 0;
  PRUint8 * textBytesCopy = NULL;
  SerialDataOutputStream textStream;

  NULLCHK( messageText );

  CNDCHK( packet == NULL, IMQ_OUT_OF_MEMORY );

  // Get the raw UTF8
  textBytes = messageText->getBytes();
  textBytesSize = messageText->getBytesSize();
  ASSERT( (textBytes != NULL) || (textBytesSize == 0) );
  ASSERT( textBytesSize >= 0 );

  // Copy the bytes
  MEMCHK( textBytesCopy = new PRUint8[textBytesSize] );
  memcpy( textBytesCopy, textBytes, textBytesSize );
  textBytes = NULL;

  // Set the message body.
  packet->setMessageBody(textBytesCopy, textBytesSize);
  textBytesCopy = NULL;

  return IMQ_SUCCESS;
Cleanup:

  DELETE_ARR( textBytesCopy );
  return errorCode;
}


/*
 *
 */
iMQError
TextMessage::getMessageText(UTF8String ** const messageText)
{
  CHECK_OBJECT_VALIDITY();

  iMQError errorCode = IMQ_SUCCESS;
  NULLCHK( messageText );
  *messageText = NULL;

  CNDCHK( packet == NULL, IMQ_OUT_OF_MEMORY );
  MEMCHK( *messageText = new UTF8String((const char*)packet->getMessageBody(), 
                                        packet->getMessageBodySize()) );
  
  return IMQ_SUCCESS;
Cleanup:
  DELETE( *messageText );
  return errorCode;
}

/*
 *
 */
iMQError 
TextMessage::getMessageTextString(const char ** const messageText)
{
  CHECK_OBJECT_VALIDITY();

  iMQError errorCode = IMQ_SUCCESS;

  NULLCHK( messageText );
  *messageText = NULL;
  CNDCHK( packet == NULL, IMQ_OUT_OF_MEMORY );
  
  // Delete the current copy of the message body
  DELETE( messageBodyText );

  // Allocate a new string for the message body
  MEMCHK( messageBodyText = new UTF8String((const char*)packet->getMessageBody(), 
                                           packet->getMessageBodySize()) );
  
  // Return the output string
  *messageText = messageBodyText->getCharStr();

  return IMQ_SUCCESS;
Cleanup:
  return errorCode;
}

/*
 *
 */
iMQError 
TextMessage::setMessageTextString(const char * const messageText)
{
  CHECK_OBJECT_VALIDITY();

  iMQError errorCode = IMQ_SUCCESS;
  const PRUint8 * textBytes = NULL;
  PRInt32 textBytesSize = 0;
  PRUint8 * textBytesCopy = NULL;
  SerialDataOutputStream textStream;

  NULLCHK( messageText );
  CNDCHK( packet == NULL, IMQ_OUT_OF_MEMORY );

  // Get the raw UTF8
  textBytes = (const PRUint8*) messageText;
  textBytesSize = (PRInt32)STRLEN( messageText );
  ASSERT( (textBytes != NULL) || (textBytesSize == 0) );
  ASSERT( textBytesSize >= 0 );

  // Copy the bytes
  MEMCHK( textBytesCopy = new PRUint8[textBytesSize] );
  memcpy( textBytesCopy, textBytes, textBytesSize );
  textBytes = NULL;

  // Set the message body.
  packet->setMessageBody(textBytesCopy, textBytesSize);
  textBytesCopy = NULL;

  return IMQ_SUCCESS;
Cleanup:

  DELETE_ARR( textBytesCopy );
  return errorCode;
}


/*
 *
 */
HandledObjectType
TextMessage::getObjectType() const 
{
  CHECK_OBJECT_VALIDITY();

  return TEXT_MESSAGE_OBJECT;
}



/*
 *
 */
HandledObjectType
TextMessage::getSuperObjectType() const
{
  CHECK_OBJECT_VALIDITY();

  return MESSAGE_OBJECT;
}

